src/domain.py

```diff
@@ -357,3 +361,70 @@ def build_prometheus_config(charm_config):
 
     logger.debug("Build prom config: {}".format(prometheus_config))
     return prometheus_config
+
+
+def _prometheus_http_api_call(
+        model_name, app_name, method, endpoint, return_response=True):
+    host = "{0}.{1}.svc:{2}".format(
+        app_name, model_name, PROMETHEUS_ADVERTISED_PORT
+    )
+    conn = http.client.HTTPConnection(host)
+    conn.request(method=method, url=endpoint)
+    logger.debug("Calling Prom API: {0} {1}".format(method, endpoint))
+
+    # TODO: Handle the un-available API endpoint case
+    response = conn.getresponse()
+    if response.status != 200:
+        logger.error("API returned error: {0}".format(
+            response.read()
+        ))
+        raise PrometheusAPIError("Reload API returned error")
+
+    if return_response:
+        try:
+            return json.loads(response.read())
+        except (ValueError, TypeError) as e:
+            logger.error("Prom API returned non-JSON: {0}".format(e))
+            raise PrometheusAPIError("Non-JSON response returned")
```

Comment:
Since this is a generic http api call function, shouldn't we not care about the return value and just return the response directly to the caller? Also take note of the `raise PrometheusAPIError("Reload API returned error")` above. This method is not being used exclusively for calling `/-/reload`.

Reply:
I think since almost all of the APIs are containing JSON response, we should take care about un-marshalling the JSON string received into the python dict in the single place, rather than having `json.loads` invocation across the whole charm.

Re: wording - good catch, fixed it.